src_kt/java/arcs/storage/DirectStore.kt

```diff
@@ -0,0 +1,410 @@
+import arcs.crdt.CrdtModel
+import arcs.crdt.CrdtModelType
+import arcs.crdt.CrdtOperation
+import arcs.storage.DirectStore.State.Name.AwaitingDriverModel
+import arcs.storage.DirectStore.State.Name.AwaitingResponse
+import arcs.storage.DirectStore.State.Name.Idle
+import kotlinx.atomicfu.AtomicRef
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.getAndUpdate
+import kotlinx.atomicfu.update
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.coroutineContext
+
+/** An [ActiveStore] capable of communicating directly with a [Driver]. */
+// TODO: generics here are sub-optimal, can we make this class generic itself?
+class DirectStore internal constructor(
+  options: StoreOptions<CrdtData, CrdtOperation, Any?>,
+  internal val localModel: CrdtModel<CrdtData, CrdtOperation, Any?>,
+  internal val driver: Driver<CrdtData>
+) : ActiveStore<CrdtData, CrdtOperation, Any?>(options) {
+  override val versionToken: String?
+    get() = driver.token
+
+  override val localData: CrdtData
+    get() = synchronized(this) { localModel.data }
+
+  /**
+   * [AtomicRef] of a [CompletableDeferred] which will be completed when the [DirectStore]
+   * transitions into the Idle state.
+   */
+  private val idleDeferred: AtomicRef<IdleDeferred> = atomic(CompletableDeferred(Unit))
+  /**
+   * [AtomicRef] of a list of [PendingDriverModel]s, allowing us to treat it as a copy-on-write,
+   * threadsafe list using [AtomicRef.update].
+   */
+  private var pendingDriverModels = atomic(listOf<PendingDriverModel>())
+  private var version = atomic(0)
+  private var state: AtomicRef<State> = atomic(State.Idle(idleDeferred, driver))
+  private val nextCallbackToken = atomic(1)
+  private val callbacks = atomic(mapOf<Int, ProxyCallback<CrdtData, CrdtOperation, Any?>>())
+
+  override suspend fun idle() = state.value.idle()
+
+  override fun on(callback: ProxyCallback<CrdtData, CrdtOperation, Any?>): Int {
+    val token = nextCallbackToken.getAndIncrement()
+    callbacks.update { callbackMap ->
+      return@update callbackMap + (token to callback)
+    }
+    return token
+  }
+
+  override fun off(callbackToken: Int) {
+    callbacks.update { callbackMap ->
+      return@update callbackMap - callbackToken
+    }
+  }
+
+  /**
+   * Receives operations/model-updates from connected storage proxies.
+   *
+   * Additionally, StorageProxy objects may request a SyncRequest, which will result in an
+   * up-to-date model being sent back to that StorageProxy. a return value of true implies that the
+   * message was accepted, a return value of false requires that the proxy send a model sync.
+   */
+  override suspend fun onProxyMessage(
+    message: ProxyMessage<CrdtData, CrdtOperation, Any?>
+  ): Boolean {
+    return when (message) {
```

Comment:
Yeah, it could, but it's still king of up in the air WRT whether or not it's okay to have a function expression span a bunch of lines like that as far as Google's Kotlin Style goes.